presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -696,22 +697,28 @@ protected ExtractionResult visitInPredicate(InPredicate node, Boolean complement
 
             InListExpression valueList = (InListExpression) node.getValueList();
             checkState(!valueList.getValues().isEmpty(), "InListExpression should never be empty");
-
-            ImmutableList.Builder<Expression> disjuncts = ImmutableList.builder();
-            for (Expression expression : valueList.getValues()) {
-                disjuncts.add(new ComparisonExpression(EQUAL, node.getValue(), expression));
+            Iterator<Expression> expressionIterator = valueList.getValues().iterator();
+            TupleDomain<Symbol> tupleDomain = process(new ComparisonExpression(EQUAL, node.getValue(), expressionIterator.next()), complement)
```

Comment:
start with `TupleDomain.all()` or `TupleDomain.none()` for complement

Reply:
Then you should be able to get away without the iterator. Also, it may be good to separate `complement` and `!complement` cases. Something like this.
```
if (complement) {
    // NOT IN
    // ...

} else {
    // IN
    TupleDomain<Symbol> tupleDomain = TupleDomain.none();
    for (Expression expression : valueList.getValues()) {
        ExtractionResult extractionResult = process(expression, complement);
        tupleDomain = ...

        if (tupleDomain.isAll() || !extractionResult.getRemainingExpression().equals(TRUE_LITERAL)) {
            // preserve original IN predicate as remaining predicate
            return new ExtractionResult(tupleDomain, node);
        }
    }
    return new ExtractionResult(tupleDomain, TRUE_LITERAL);
}
```

Reply:
I initially thought an approach like this. Will implement in this model .